### ISSUE TITLE
Name constraints are being checked for certificates that specify them

### ISSUE THREAD

**Issue by robertsuska** [randombit/botan#2737]
Hi,

I've noticed that when Botan checks name constraints during validation of X.509 certificates, it checks them also for certificates in which they are specified. I think this might be a bug since [RFC 5280](https://www.rfc-editor.org/rfc/rfc5280.txt) (article 4.2.1.10) says. _"The name constraints extension, which MUST be used only in a CA certificate, indicates a name space within which all subject names in **subsequent** certificates in a certification path MUST be located."_ (emphasis mine).

I've included an [example_chain_error.zip](https://github.com/randombit/botan/files/6421097/example_chain_error.zip), where

_root_ca.pem_ is `Subject: C = CZ, L = Prague, O = TestRootCA`

_intermediate_ca_error.pem_ is `C = CZ, L = Prague, O = TestIntermediateCA` that specifies a single name constraint (but does not comply with it):

```
X509v3 Name Constraints: critical
    Permitted:
      DirName:C = CZ, L = Brno
```

 _server.pem_ is `Subject: C = CZ, L = Brno, O = TestServer` which complies with the given name constraint

and validation of this chain (_root_ca.pem_ -> _intermediate_ca_error.pem_ -> _server.pem_) returns code 4004 (NAME_CONSTRAINT_ERROR).

I've also included [example_chain_ok.zip](https://github.com/randombit/botan/files/6421025/example_chain_ok.zip), where _intermediate_ca_ok.pem_ is `C = CZ, L = Brno, O = TestIntermediateCA` and so it does comply with its name contraint, and the validation returns code 0 (VERIFIED).

---

I am using Botan in version 2.17.2 through its Python binding.

The two chains can be validated for example like this:

```
import botan2 as botan

print("example chain without violation: ", end="")
print(botan.X509Cert(filename="example_chain_ok/server.pem").verify(trusted_path="example_chain_ok/trust_store/",
                                                                    intermediates=[botan.X509Cert(filename="example_chain_ok/intermediate_ca_ok.pem")],
                                                                    reference_time=1622394060))
print("example chain with violation: ", end="")
print(botan.X509Cert(filename="example_chain_error/server.pem").verify(trusted_path="example_chain_error/trust_store/",
                                                                    intermediates=[botan.X509Cert(filename="example_chain_error/intermediate_ca_error.pem")],
                                                                    reference_time=1622394060))
```

I've tested this also with other libraries, namely OpenSSL, pyOpenSSL, GnuTLS, and MbedTLS, and none of them behave the same way.

**Comment by randombit** [comment]
There was logic to avoid checking self but it for whatever reason was written to only do so when it was the root CA which had the name constraint, not intermediates. I'm not sure why that was.

Thanks for the report and test cases by the way, much appreciated.

**Comment by randombit** [comment]
Now fixed in master and 2.18.1
